### PR TITLE
Add support for 'console' output type of CodeNarc plugin (Fix 1481)

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

See #1481 and [discussion in old forum](https://discuss.gradle.org/t/codenarc-reportformat-console-still-hides-the-output-from-the-console/11971/2). 

Although it is documented that CodeNarc plugin supports 'console' output type, that type is never implemented. This commit add 'console' TaskGeneratedSingleFile to CodeNarcReportsImpl, and adjust the options with which Ant codenarc task is called. To redirect Ant output to Gradle standard output in 'console' type, the lifecycleLogLevel of AntLogAdapter is changed to INFO.

Then, a unit test and an integration test are provided.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] [Sign Gradle CLA](http://gradle.org/contributor-license-agreement/)
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Ensure that tests pass locally: `./gradlew quickCheck <impacted-subproject>:check`

### CODE REVIEW

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy [hunk 1]
```diff
@@ -73,4 +82,12 @@ abstract class CodeNarcInvoker {
         }
     }
 
+    static void setLifecycleLogLevel(Object ant, String lifecycleLogLevel) {
+        ant?.builder?.project?.buildListeners?.each {
+            // We cannot use instanceof or getClass()==AntLoggingAdapter since they're in different class loaders
+            if (it.class.simpleName == 'AntLoggingAdapter') {
```
Comment (jjohannes) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy]: Looks like there is no simpler way to do this.

We can improve it slightly, by replacing the check with
`it.class.name == AntLoggingAdapter.name`

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy [hunk 2]
```diff
@@ -47,8 +47,17 @@ abstract class CodeNarcInvoker {
             try {
                 ant.codenarc(ruleSetFiles: "file:${configFile}", maxPriority1Violations: maxPriority1Violations, maxPriority2Violations: maxPriority2Violations, maxPriority3Violations: maxPriority3Violations) {
                     reports.enabled.each { SingleFileReport r ->
-                        report(type: r.name) {
-                            option(name: 'outputFile', value: r.destination)
+                        // See http://codenarc.sourceforge.net/codenarc-TextReportWriter.html
+                        if (r.name == 'console') {
+                            setLifecycleLogLevel(ant,'INFO')
```
Comment (jjohannes) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy]: Following Gradle's code style rules, this should be `(ant, 'INFO')` <- with a _space_

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy [hunk 3]
```diff
@@ -47,8 +47,17 @@ abstract class CodeNarcInvoker {
...
+                            report(type: 'text') {
+                                option(name: 'writeToStandardOut', value: true)
+                            }
+                        } else {
+                            setLifecycleLogLevel(ant,null)
```
Comment (jjohannes) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy]: `(ant, null)`

**Comment by jjohannes** [comment]
@blindpirate PR looks good, thanks! Just some minor remarks. Could you address them? Then this can go into 4.1

I think this also deserves a short mentioning in the Release Notes. Could you add a `###` section with one or two sentences in the `## New and noteworthy` section as part of the PR?
[subprojects/docs/src/docs/release/notes.md](
https://github.com/gradle/gradle/blob/983212247682ba92ad03328abce8dc24c04eab51/subprojects/docs/src/docs/release/notes.md)
Don't worry too much about wording. I'll polish it if needed.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy (hunk 1 above)
Comment (blindpirate) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy]: Yes, you're right.

File: subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy (hunk 2 above)
Comment (blindpirate) [inline subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/CodeNarcInvoker.groovy]: Done. You're so careful.

**Comment by blindpirate** [comment]
As you say, I have improved it. Please let me know if there're any other questions.

**Comment by jjohannes** [comment]
Thanks @blindpirate! The PR is merged and will be part of the 4.1 release.

**Comment by blindpirate** [comment]
Thanks @jjohannes .

**Comment by vgaidarji** [comment]
@blindpirate did you consider implementing support for multiple reports?
It might be useful to have both 'console' and 'html' reports generated, so that some people can check console for errors, others will open nice visual HTML report somewhere in CI.
How difficult it will be to support multiple reports?

**Comment by blindpirate** [comment]
@vgaidarji Right now I don't think have enough bandwidth/priority to implement such a feature. However, I think this shouldn't be too difficult, maybe at most 100 lines change? We're glad if any contributions can be made.

**Comment by twilco** [comment]
So just to confirm, multiple report format outputs aren't currently supported with this plugin?  I have to choose either console, or HTML, or XML?

**Pull request merged**